### PR TITLE
Generate empty flag enums

### PR DESCRIPTION
**Pull request by jobh**
We should sample empty enum.Flag examples by default. They are valid instances after all, representing no flag bits set.

True, a user could opt-in by for example `one_of(sampled_from(Flags), just(Flags(0)))`. But it's better to have complete coverage by default, and if necessary opt-out by adding `.filter(bool)` or similar.

We *might* want to lessen visible impact by shrinking towards first non-empty flag instead of empty. I don't know how best to achieve this, I expected the first (opt-in) example to do exactly that but it doesn't. Anyway, I added a provisional test to modify to whatever shrinking behaviour we desire.

### CODE REVIEW

**Comment by jobh** [comment]
An additional idea:

Subsets (`sets(sampled_from(...))`) - including subsets of flags - I believe they skew towards shorter lengths.

This is natural for constructive sets (e.g. `sets(integers())`), but possibly sampled sets (subsets) should have a higher probability of generating complete or close-to-complete sets.

Just a though. For a another day.

**Comment by Zac-HD** [comment]
> Subsets (`sets(sampled_from(...))`) - including subsets of flags - I believe they skew towards shorter lengths.
> 
> This is natural for constructive sets (e.g. `sets(integers())`), but possibly sampled sets (subsets) should have a higher probability of generating complete or close-to-complete sets.

Hmm, yeah, this makes sense to me.  How about something like:

```python
empty = elements(0)
inner = st.tuples(*[st.sampled_from([elem, empty]) for elem in elements]).map(
    # Shrink to just-the-first by using xor to invert that option.  len-zero special-cased in sampled_from()
    lambda s: reduce(operator.xor, s, next(iter(elements)))
)
```

We'll want to add a test to sample from an `enum.Flag` of length zero, and add a `if len(elements) == 0 and isinstance(elements, Flag): return just(elements(0))` special case in the enum-specific error branch which it currently triggers.

Plus a release note, for a minor release :slight_smile:

**Comment by jobh** [comment]
> ```python
> st.tuples(*[st.sampled_from([elem, empty]) for elem in elements]).map(
>     # Shrink to just-the-first by using xor to invert that option.  len-zero special-cased in sampled_from()
>     lambda s: reduce(operator.xor, s, next(iter(elements)))
> ```

`xor` doesn't really work well; the order of empty and first flag is exchanged everywhere, so it doesn't shrink to any other single flag bit. I ended up with a split strategy using `one_of`, where the first branch just reorders empty wrt named flags and the second branch excludes empty; but ensuring that the two branches shrink in the same way.

I like your distribution though, although I couldn't get it _just_ right - `sampled_from([elem, empty])` shrinks the wrong way, while `sampled_from([empty, elem])` doesn't generate any all-`elem` samples which I'm looking for.

Anyhow: I added tests and release note, so mostly done pending test runs and review.

File: hypothesis-python/src/hypothesis/strategies/_internal/core.py
Comment (jobh) [inline hypothesis-python/src/hypothesis/strategies/_internal/core.py:212]: Moved this before `len(values)` check, as `enum.Flag` always have at least one valid value (empty).

**Comment by jobh** [comment]
I see there are more quirks to consider. I guess I'll put this PR on hold for a little while until this is sorted.

```
# The following is a TypeError on py 3.12, despite the exception msg explicitly saying
# this is how to construct an empty Flag. We can't do anything about this, but we can
# raise InvalidArgument instead of letting the TypeError through.
>>> A = enum.Flag("A", names=()); A(0), list(A)
(<A: 0>, [])

# Named empty seems to be accepted on 3.12, too. In this case, perhaps we
# should shrink to it since it's named.
>>> A = enum.Flag("A", names={"a": 0}); A(0), list(A)
(<A.a: 0>, [])

# Composite flag values are not visible in values list, so we won't generate them.
# Perhaps we should, or just warn/raise when we detect composite values without
# corresponding single-bit values?
>>> A = enum.Flag("A", names={"a": 3}); A(0), list(A)
(<A: 0>, [])
```

**Comment by Zac-HD** [comment]
Oh wow, seems like we might have a CPython bug report on our hands again...

**Comment by jobh** [comment]
> Oh wow, seems like we might have a CPython bug report on our hands again...

I think so, at least the error message seems misleading. But I haven't close-read checked it yet. Just quick clarification, "until this is sorted" was meant as "until I sort it out", nothing else ;-) I think I have a handle on it now, just need to clean up and fix remaining tests.

**Comment by jobh** [comment]
> > Oh wow, seems like we might have a CPython bug report on our hands again...
> 
> I think so, at least the error message seems misleading. But I haven't close-read checked it yet. Just a quick clarification: "until this is sorted" was meant as "until I sort it out", nothing more sinister ;-)

I think I have a handle on it now, just need to clean up and fix remaining tests.

**Comment by jobh** [comment]
> > Oh wow, seems like we might have a CPython bug report on our hands again...
> 
> I think so, at least the error message seems misleading. But I haven't close-read checked it yet. Just a quick 

Yes, it looks unambiguous. https://github.com/python/cpython/issues/114474

**Comment by jobh** [comment]
Ready for proper review now @Zac-HD ! I know about the RELEASE xref thing, I'll fix that later, don't want to abort the ongoing tests.

File: hypothesis-python/tests/conjecture/test_float_encoding.py
Comment (jobh) [inline hypothesis-python/tests/conjecture/test_float_encoding.py:123]: I stole the implementation in `compat` from here, so this change is just to avoid duplication. And get some coverage for free.

File: hypothesis-python/tests/cover/test_sampled_from.py
Comment (jobh) [inline hypothesis-python/tests/cover/test_sampled_from.py:67]: After the latest commit, I don't think this test is necessary for coverage anymore - it just defers to the generic sample-empty-sequence error. Since the future cpy behaviour is not certain, should I just drop this test?

**Comment by jobh** [comment]
New flakiness:
```
FAILED tests/cover/test_lookup.py::test_resolves_flag_enum[sampled_from] - hypothesis.errors.FailedHealthCheck: Examples routinely exceeded the max allowable size. (20 examples overran while generating 8 valid ones)...
```

This test filters out single-flag values from `sampled_from(Flag)`, and the sample distribution has obviously changed. Should I change the test (to use `find_any` for example), or does "overran" imply something worse?

File: hypothesis-python/tests/cover/test_sampled_from.py
Comment (jobh) [reply hypothesis-python/tests/cover/test_sampled_from.py]: I dropped it

**Comment by jobh** [comment]
The remaining test failure looks unrelated. It may or may not be relevant that it turned up after I rebased to latest master to fix a merge conflict.

https://github.com/HypothesisWorks/hypothesis/actions/runs/7625246397/job/20769107602?pr=3852

File: hypothesis-python/tests/cover/test_sampled_from.py
Comment (Zac-HD) [inline hypothesis-python/tests/cover/test_sampled_from.py]: I'd reach for our `assert_all_examples(sampled_from(enum_class), lambda x: isinstance(x, enum_class))` test helper here.

**Review by Zac-HD — APPROVED** [review]
Added one more testcase, a Flag with no values and an unnamed empty.  Looks great and I'm excited to merge it!  

The flaky test was my fault from https://github.com/HypothesisWorks/hypothesis/pull/3839; the pre-existing test picked up a rare failure mode if it generated a non-ascii literal _and_ the re.ASCII flag together.  Checking that case and disabling the flag restores the stability of the test under our new-ish semantics.

**Comment by jobh** [comment]
Before we merge this one - I didn't see it lately but there was a flake at: https://github.com/HypothesisWorks/hypothesis/actions/runs/7623512734/job/20763710693#step:6:778

If it's not serious, I'll just drop the filter from this test since we already check combination coverage in nocover/test_sampled_from. Or is the overrun indicative of something bad?

**Comment by jobh** [comment]
> Added one more testcase, a Flag with no values and an unnamed empty. Looks great and I'm excited to merge it!

Thanks! Note, the new test case runs straight into the recent python TypeError-on-empty (and hence no constructable instances to sample from)

**Review by Zac-HD — APPROVED** [review]
🎉 

Thanks again @jobh, glad to have this feature, and more broadly to have you contributing regularly 😍

**Pull request merged**
